Python/pyhash.c

```diff
@@ -131,24 +131,23 @@ _Py_HashDouble(PyObject *inst, double v)
     return (Py_hash_t)x;
 }
 
+// See also _Py_hashtable_hash_ptr() which is similar but can return -1.
 Py_hash_t
-_Py_HashPointerRaw(const void *p)
+PyHash_Pointer(const void *ptr)
 {
-    size_t y = (size_t)p;
-    /* bottom 3 or 4 bits are likely to be 0; rotate y by 4 to avoid
-       excessive hash collisions for dicts and sets */
-    y = (y >> 4) | (y << (8 * SIZEOF_VOID_P - 4));
-    return (Py_hash_t)y;
-}
+    uintptr_t x = (uintptr_t)ptr;
+    Py_BUILD_ASSERT(sizeof(x) == sizeof(ptr));
 
-Py_hash_t
-_Py_HashPointer(const void *p)
-{
-    Py_hash_t x = _Py_HashPointerRaw(p);
-    if (x == -1) {
-        x = -2;
+    // Bottom 3 or 4 bits are likely to be 0; rotate x by 4 to the right
+    // to avoid excessive hash collisions for dicts and sets.
+    x = (x >> 4) | (x << (8 * SIZEOF_UINTPTR_T - 4));
```

Comment:
Why do you duplicate the code from `_Py_hashtable_hash_ptr`?

Reply:
They're different, but all the code above the `if` could be replaced by `Py_hash_t result = _Py_hashtable_hash_ptr(ptr);` to avoid the duplication.

If they need to be separate, would it work to add the common implementation as a `static inline` in `pycore_pyhash.h`?

Reply:
Ok, I kept _Py_HashPointerRaw() as a static inline function to avoid code duplication.